File: core/src/main/java/vn/loitp/views/animation/flyschool/FlyPath.java
package vn.loitp.views.animation.flyschool;

import android.graphics.Path;

public abstract class FlyPath {
    public static SingleLinePath getSimpleLinePath(FPoint mFPoint) {
        return new SingleLinePath(mFPoint);
    }

    public static BeizerPath getBeizerPath(FPoint mFPoint1, FPoint mFPoint2, FPoint mFPoint3) {
        return new BeizerPath(mFPoint1, mFPoint2, mFPoint3);
    }

    public static MultipleLinePath getMultipleLinePath(FPoint... mFPoints) {
        if (mFPoints == null) {
            throw new IllegalArgumentException("Passed FPoints should not be null");
        }

        MultipleLinePath multipleLinePath = null;
        if (mFPoints.length > 0) {
            multipleLinePath = new MultipleLinePath();
            for (FPoint fPoint : mFPoints) {
                multipleLinePath.addFPoint(fPoint);
            }
        }

        return multipleLinePath;

    }

    /**
     * @param mOrigin : Origin point where the PATH should start
     * @param width   : Width of the {@link ShapeFlyer} in pixels
     * @param height  : Height of the {@link ShapeFlyer} in pixels
     *                <p>
     *                Should return the path/trajectory of the animation.
     *                Please refer to {@link BeizerPath} and {@link SingleLinePath} for examples
     */
    public abstract Path getPath(FPoint mOrigin, float width, float height);
}

File: core/src/main/java/com/loitp/anim/flySchool/FlyPath.kt
package com.loitp.anim.flySchool

import android.graphics.Path

abstract class FlyPath {
    /**
     * @param mOrigin : Origin point where the PATH should start
     * @param width   : Width of the [ShapeFlyer] in pixels
     * @param height  : Height of the [ShapeFlyer] in pixels
     *
     *
     * Should return the path/trajectory of the animation.
     * Please refer to [BeizerPath] and [SingleLinePath] for examples
     */
    abstract fun getPath(
        mOrigin: FPoint,
        width: Float,
        height: Float
    ): Path

    companion object {
        fun getSimpleLinePath(mFPoint: FPoint): SingleLinePath {
            return SingleLinePath(mFPoint)
        }

        fun getBeizerPath(
            mFPoint1: FPoint,
            mFPoint2: FPoint,
            mFPoint3: FPoint
        ): BeizerPath {
            return BeizerPath(mFPoint1, mFPoint2, mFPoint3)
        }

        @Suppress("unused")
        fun getMultipleLinePath(vararg mFPoints: FPoint?): MultipleLinePath? {
            var multipleLinePath: MultipleLinePath? = null
            if (mFPoints.isNotEmpty()) {
                multipleLinePath = MultipleLinePath()
                for (fPoint in mFPoints) {
                    multipleLinePath.addFPoint(fPoint)
                }
            }
            return multipleLinePath
        }
    }
}
